### PR TITLE
Add parameter eagerClose to Flow.interleave #22710

### PR DESCRIPTION
**Pull request by evis**

### CODE REVIEW

**Comment by ktoso** [comment]
OK TO TEST

**Comment by ktoso** [comment]
Refs  #22710

File: akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala]: This is not binary compatible, it must be done as overload of the method, so:

```
def interleave[U >: Out](that: Graph[SourceShape[U], _], segmentSize: Int): Repr[U] = 
  interleave(that, segmentSize, eagerClose = false)

def interleave[U >: Out](that: Graph[SourceShape[U], _], segmentSize: Int, eagerClose: Boolean): Repr[U] = ???
```

File: akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala]: (same here)

**Comment by evere10** [comment]
Is it ok to copy-paste javadocs between overloaded methods or it's better to just mention one in another's docs?

**Comment by evere10** [comment]
Why does Jenkins PR Validation shows "Test FAILed", if failed tests count is zero?

**Comment by patriknw** [comment]
It's binary compatibility checks that fails:
```
[info] akka-stream: found 8 potential binary incompatibilities while checking against com.typesafe.akka:akka-stream_2.11:2.4.4  (filtered 464)
[error]  * method interleaveGraph(akka.stream.Graph,Int)akka.stream.Graph in class akka.stream.scaladsl.GraphDSL#Implicits#PortOpsImpl does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("akka.stream.scaladsl.GraphDSL#Implicits#PortOpsImpl.interleaveGraph")
[error]  * method interleaveMat(akka.stream.Graph,Int,Boolean,scala.Function2)akka.stream.scaladsl.FlowOpsMat in trait akka.stream.scaladsl.FlowOpsMat is present only in current version
[error]    filter with: ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.stream.scaladsl.FlowOpsMat.interleaveMat")
[error]  * method interleaveGraph(akka.stream.Graph,Int)akka.stream.Graph in class akka.stream.scaladsl.Flow does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("akka.stream.scaladsl.Flow.interleaveGraph")
[error]  * method interleaveGraph(akka.stream.Graph,Int)akka.stream.Graph in class akka.stream.scaladsl.Source does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("akka.stream.scaladsl.Source.interleaveGraph")
[error]  * method interleaveGraph(akka.stream.Graph,Int)akka.stream.Graph in trait akka.stream.scaladsl.FlowOps does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("akka.stream.scaladsl.FlowOps.interleaveGraph")
[error]  * method interleaveGraph(akka.stream.Graph,Int,Boolean)akka.stream.Graph in trait akka.stream.scaladsl.FlowOps is present only in current version
[error]    filter with: ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.stream.scaladsl.FlowOps.interleaveGraph")
[error]  * synthetic method interleaveGraph$default$3()Boolean in trait akka.stream.scaladsl.FlowOps is present only in current version
[error]    filter with: ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.stream.scaladsl.FlowOps.interleaveGraph$default$3")
[error]  * method interleave(akka.stream.Graph,Int,Boolean)akka.stream.scaladsl.FlowOps in trait akka.stream.scaladsl.FlowOps is present only in current version
[error]    filter with: ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.stream.scaladsl.FlowOps.interleave")
```

I think those should be fine to add as exceptions to MiMa.scala, but we have to look at it when reviewing.

File: akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala]: I'd make this:
"If `eagerClose` is `false` and one of the upstreams complete the elements from the other upstream will continue passing through the interleave stage. If `eagerClose` is `true` and one of the upstream complete interleave will cancel the other upstream and complete itself"

I'd also skip the source samples, since I think the description is enough and code in scaladoc is hard to maintain as it does not get compiled and tested.

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowInterleaveSpec.scala
Comment (johanandren) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowInterleaveSpec.scala]: Would be good to also have test coverage where:
* the other stream closes (probably no issues here, but for regression coverage)
* the source that completes does not do it at a `segmentSize` boundary

**Review by johanandren — COMMENTED** [review]
Looking good, sorry that it took so long to get reviewed. I added some notes of possible improvements.

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowInterleaveSpec.scala
Comment (evis) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowInterleaveSpec.scala]: > the source that completes does not do it at a segmentSize boundary

What do you mean here? Why shouldn't it complete at a segmentSize boundary? Do you mean, that such code:

`Source(0 to 2).interleave(Source(3 to 5), 3, eagerClose = true)`

shouldn't complete at `Source(0 to 2)` and should return `Source(0 to 5)` instead?

**Comment by lightbend-cla-validator** [comment]
Hi @evis,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Lightbend Contributors License Agreement:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

**Comment by evis** [comment]
@johanandren could you review my recent commits?

**Comment by evis** [comment]
@johanandren @ktoso ping for review =)

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowInterleaveSpec.scala
Comment (johanandren) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowInterleaveSpec.scala]: I didn't mean that the logic is wrong but that it would be good to have test coverage for both the case where segment size is a multiple of the number of elements in the completing source and where it is not. `0 to 2` is `0, 1, 2` so it already isn't on the boundary, but `0 to 1` would be.

And then a left hand `Source` that continues but the passed in `Source` completes early.

**Comment by johanandren** [comment]
Sorry about slow reviewing here, we have been (and are) busy (and vacations, travels and conference talks).

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowInterleaveSpec.scala
Comment (johanandren) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowInterleaveSpec.scala]: Actually, this was just about improving the tests, you can just skip that.

File: project/MiMa.scala
Comment (johanandren) [inline project/MiMa.scala]: As we did a few releases while this PR was already here this needs to be rebased and moved to a new "2.5.3" map entry, and the set of excludes it should only occur once, not multiple times.

**Review by johanandren — APPROVED** [review]
LGTM with the MiMa filters fixed (maybe we could do that and merge it if we can get another core team LGTM)

File: project/MiMa.scala
Comment (evis) [inline project/MiMa.scala]: Moved filters from 2.5.1 and 2.5.2 to 2.5.3, sorry for delay.

**Comment by evis** [comment]
@ktoso could you give LGTM too? =)

**Review by ktoso — APPROVED** [review]
LGTM, looking good and binary compatible :)

Thanks and sorry so much for the delays

**Comment by ktoso** [comment]
The merge conflict is a bit weird here, I'll take care of it.

**Comment by patriknw** [comment]
I have fixed the merge conflict in https://github.com/akka/akka/pull/23718 (to be merged after validation). Thanks @evis !

**Pull request closed without merging**
